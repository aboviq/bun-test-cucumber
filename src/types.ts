import type { ExpressionFactory } from '@cucumber/cucumber-expressions';
import parseTags from '@cucumber/tag-expressions';

export type TagExpression = ReturnType<typeof parseTags>;

export type CucumberExpression = ReturnType<ExpressionFactory['createExpression']>;

type GherkinDataType =
  | 'int'
  | 'float'
  | 'double'
  | 'word'
  | 'string'
  | 'bigdecimal'
  | 'biginteger'
  | 'byte'
  | 'short'
  | 'long'
  // anonymous
  | '';

type GherkinDataTypeToType<T extends GherkinDataType> = {
  int: number;
  float: number;
  double: number;
  word: string;
  string: string;
  bigdecimal: string;
  biginteger: bigint;
  byte: number;
  short: number;
  long: number;
  '': string;
}[T];

type GherkinDataTypesToTypes<T extends GherkinDataType[]> = {
  [K in keyof T]: T[K] extends GherkinDataType ? GherkinDataTypeToType<T[K]> : never;
};

type ExtractGherkinDataType<
  T extends string,
  S extends GherkinDataType[],
> = T extends `${string}{${infer U extends GherkinDataType}` ? [...S, U] : S;

type ExtractGherkinDataTypes<T extends string, S extends GherkinDataType[]> = T extends `${infer U}}${infer W}`
  ? ExtractGherkinDataTypes<W, ExtractGherkinDataType<U, S>>
  : ExtractGherkinDataType<T, S>;

export type ExtractTypes<T extends string> = GherkinDataTypesToTypes<ExtractGherkinDataTypes<T, []>>;

export type HookType = 'beforeAll' | 'before' | 'beforeStep' | 'afterStep' | 'after' | 'afterAll';

export type HookFunction<State> = (state: State) => State | Promise<State>;

export type Hook<State> = {
  name: string;
  fn: HookFunction<State>;
  options?: { tags?: string; tagsExpression?: TagExpression };
};

export type Step = {
  expression: string;
  cucumberExpression: CucumberExpression;
  fn: Function;
};
