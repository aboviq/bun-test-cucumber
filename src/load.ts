import { Glob } from 'bun';

/**
 * Load all feature files matching the given glob pattern
 *
 * This will import all matching files which will trigger the Bun Test Cucumber plugin
 * to process the feature files and generate the corresponding test files.
 *
 * When <https://github.com/oven-sh/bun/issues/3440> is resolved this won't be necessary.
 *
 * @param pattern Glob pattern to the feature files (relative to `cwd` if set, otherwise relative to `process.cwd()`)
 * @param cwd Optional current working directory to resolve the pattern against (defaults to process.cwd())
 */
export const loadFeatures = async (pattern: string, cwd?: string): Promise<void> => {
  const featuresGlob = new Glob(pattern);

  for await (const file of featuresGlob.scan({
    absolute: true,
    cwd,
  })) {
    await import(file);
  }
};
