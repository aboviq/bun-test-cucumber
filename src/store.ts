import { ExpressionFactory, ParameterTypeRegistry, type Argument } from '@cucumber/cucumber-expressions';
import { PickleStepType, type PickleStep, type PickleTag } from '@cucumber/messages';
import parseTags from '@cucumber/tag-expressions';

import type { Hook, HookFunction, HookType, Step } from './types';

const expressionFactory = new ExpressionFactory(new ParameterTypeRegistry());

const steps: Step[] = [];

const hooks = new Map<HookType, Hook<unknown>[]>();

export const addStep = (expression: string, fn: Function): void => {
  const cucumberExpression = expressionFactory.createExpression(expression);

  if (!fn.name) {
    Object.defineProperty(fn, 'name', { value: expression });
  }

  steps.push({ expression, cucumberExpression, fn });
};

export const addHook = (type: HookType, ...args: unknown[]): void => {
  if (!hooks.has(type)) {
    hooks.set(type, []);
  }
  let name = '';
  let options: Hook<unknown>['options'];
  let fn: HookFunction<unknown> | undefined;

  for (const element of args) {
    if (typeof element === 'string') {
      name = element;
    } else if (typeof element === 'function') {
      fn = element as HookFunction<unknown>;
    } else if (typeof element === 'object' && element !== null) {
      options = element;
      options.tagsExpression = options.tags ? parseTags(options.tags) : undefined;
    }
  }

  if (!fn) {
    throw new Error('Hook function is required');
  }

  if (name && !fn.name) {
    Object.defineProperty(fn, 'name', { value: name });
  }

  hooks.get(type)!.push({ name, fn, options });
};

export async function applyHooks(type: HookType, state: unknown, tags?: PickleTag[]): Promise<unknown> {
  const hooksForType = hooks.get(type) ?? [];
  for (const hook of hooksForType) {
    if (hook.options?.tagsExpression && tags) {
      const tagNames = tags.map((tag) => tag.name);
      if (!hook.options.tagsExpression.evaluate(tagNames)) {
        continue;
      }
    }
    state = await hook.fn(state);
  }
  return state;
}

function getStepType(pickleStep: PickleStep): string {
  switch (pickleStep.type) {
    case PickleStepType.CONTEXT:
      return 'Given';
    case PickleStepType.ACTION:
      return 'When';
    case PickleStepType.OUTCOME:
      return 'Then';
    default:
      return 'Given';
  }
}

export async function runStep<State>(pickleStep: PickleStep, state: State): Promise<State> {
  const { step, parameters } = findStepDefinition(pickleStep);

  return step.fn(state, parameters, pickleStep.argument);
}

function findStepDefinition(pickleStep: PickleStep): {
  step: Step;
  parameters: Array<Argument | null>;
} {
  const matches: Array<{ step: Step; parameters: Array<Argument | null> }> = [];

  for (const step of steps) {
    const match = step.cucumberExpression.match(pickleStep.text);

    if (match) {
      matches.push({
        step,
        parameters: match.map((argument) => argument.getValue(null)),
      });
    }
  }

  const [match, ...rest] = matches;

  if (!match) {
    throw new Error(`No step definition found for step: "${pickleStep.text}"
    Implement with the following snippet:

    ${getStepType(pickleStep)}('${escape(pickleStep.text)}', async (state) => {
      return state;
    });
  `);
  }

  if (rest.length > 0) {
    throw new Error(`Multiple step definitions found for step: "${getStepType(pickleStep)}" "${pickleStep.text}"`);
  }

  return match;
}
