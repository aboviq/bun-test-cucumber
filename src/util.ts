export const escape = (str: string): string => str.replaceAll("'", "\\'");
