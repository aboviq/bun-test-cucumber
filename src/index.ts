import { type PickleStepArgument } from '@cucumber/messages';
import { addHook, addStep } from './store';
import type { HookFunction, ExtractTypes } from './types';

export { applyHooks, runStep } from './store';
export { loadFeatures } from './load';
export { bunTestCucumber } from './plugin';

export interface AddHookFunction<S> {
  /**
   * Simple hook definition function
   *
   * Available hook types are:
   * - `BeforeAll` - runs once per feature file, before any scenarios
   * - `Before` - runs before each scenario
   * - `BeforeStep` - runs before each step in a scenario
   * - `AfterStep` - runs after each step in a scenario
   * - `After` - runs after each scenario
   * - `AfterAll` - runs once per feature file, after all scenarios
   *
   * **NOTE:** Must return the state object (and can be async). Which TypeScript will enforce (especially if using {@link withState}).
   *
   * @example
   * ```ts
   * BeforeAll(async (state) => {
   *   // setup code here
   *   return state;
   * });
   *
   * // or
   *
   * Before((state) => {
   *   // setup code here
   *   return state;
   * });
   * ```
   *
   * @param fn Hook function (can be either sync or async)
   */
  <State = S>(fn: HookFunction<State>): void;
  /**
   * Hook with name definition function
   *
   * Available hook types are:
   * - `BeforeAll` - runs once per feature file, before any scenarios
   * - `Before` - runs before each scenario
   * - `BeforeStep` - runs before each step in a scenario
   * - `AfterStep` - runs after each step in a scenario
   * - `After` - runs after each scenario
   * - `AfterAll` - runs once per feature file, after all scenarios
   *
   * **NOTE:** Must return the state object (and can be async). Which TypeScript will enforce (especially if using {@link withState}).
   *
   * @example
   * ```ts
   * BeforeAll('Initial setup', async (state) => {
   *   // setup code here
   *   return state;
   * });
   *
   * // or
   *
   * Before('Running before each scenario', (state) => {
   *   // setup code here
   *   return state;
   * });
   * ```
   *
   * @param name Name of the hook (used in stack traces)
   * @param fn Hook function (can be either sync or async)
   */
  <State = S>(name: string, fn: HookFunction<State>): void;
  /**
   * Hook with options definition function
   *
   * Available hook types are:
   * - `BeforeAll` - runs once per feature file, before any scenarios
   * - `Before` - runs before each scenario
   * - `BeforeStep` - runs before each step in a scenario
   * - `AfterStep` - runs after each step in a scenario
   * - `After` - runs after each scenario
   * - `AfterAll` - runs once per feature file, after all scenarios
   *
   * **NOTE:** Must return the state object (and can be async). Which TypeScript will enforce (especially if using {@link withState}).
   *
   * @example
   * ```ts
   * BeforeAll({ tags: '@browser' }, async (state) => {
   *   // setup browser code here
   *   return state;
   * });
   *
   * // or
   *
   * Before({ tags: '@fakeTime' }, (state) => {
   *   // setup code here, e.g. fake system time
   *   return state;
   * });
   *
   * After({ tags: '@fakeTime' }, (state) => {
   *   // teardown code here, e.g. restore system time
   *   return state;
   * });
   * ```
   *
   * @param options Hook options (currently only `tags` is supported)
   * @param fn Hook function (can be either sync or async)
   */
  <State = S>(options: { tags: string }, fn: HookFunction<State>): void;
  /**
   * Hook with name and options definition function
   *
   * Available hook types are:
   * - `BeforeAll` - runs once per feature file, before any scenarios
   * - `Before` - runs before each scenario
   * - `BeforeStep` - runs before each step in a scenario
   * - `AfterStep` - runs after each step in a scenario
   * - `After` - runs after each scenario
   * - `AfterAll` - runs once per feature file, after all scenarios
   *
   * **NOTE:** Must return the state object (and can be async). Which TypeScript will enforce (especially if using {@link withState}).
   *
   * @example
   * ```ts
   * BeforeAll('Setup browser', { tags: '@browser' }, async (state) => {
   *   // setup browser code here
   *   return state;
   * });
   *
   * // or
   *
   * Before('Fake time', { tags: '@fakeTime' }, (state) => {
   *   // setup code here, e.g. fake system time
   *   return state;
   * });
   *
   * After('Restore time', { tags: '@fakeTime' }, (state) => {
   *   // teardown code here, e.g. restore system time
   *   return state;
   * });
   * ```
   *
   * @param name Name of the hook (used in stack traces)
   * @param options Hook options (currently only `tags` is supported)
   * @param fn Hook function (can be either sync or async)
   */
  <State = S>(name: string, options: { tags: string }, fn: HookFunction<State>): void;
}

export interface AddStepFunction<S> {
  <Pattern extends string, State = S>(
    pattern: Pattern,
    fn: (state: State, args: ExtractTypes<Pattern>, data?: PickleStepArgument) => State | Promise<State>,
  ): void;
}

export interface AddStepFunctionWithState<State> {
  <Pattern extends string>(
    pattern: Pattern,
    fn: (state: State, args: ExtractTypes<Pattern>, data?: PickleStepArgument) => State | Promise<State>,
  ): void;
}

export interface WithState<S> {
  BeforeAll: AddHookFunction<S>;
  Before: AddHookFunction<S>;
  BeforeStep: AddHookFunction<S>;
  AfterStep: AddHookFunction<S>;
  After: AddHookFunction<S>;
  AfterAll: AddHookFunction<S>;
  Given: AddStepFunctionWithState<S>;
  When: AddStepFunctionWithState<S>;
  Then: AddStepFunctionWithState<S>;
}

/**
 * Helper function for less repetition and better type inference
 *
 * @example
 *
 * ### Without `withState`:
 * ```ts
 * import { Before, Given } from '@aboviq/bun-test-cucumber';
 *
 * interface State {
 *  counter: number;
 * }
 *
 * Before<State>(async (state) => {
 *   return { ...state, counter: 0 };
 * });
 * // or
 * Before(async (state: State) => {
 *   return { ...state, counter: 0 };
 * });
 *
 * Given('the counter is {int}', async (state: State, [value]) => {
 *   return { ...state, counter: value };
 * });
 * ```
 *
 * ### With `withState`:
 *
 * ```ts
 * import { withState } from '@aboviq/bun-test-cucumber';
 *
 * interface State {
 *   counter: number;
 * }
 *
 * const { Before, Given } = withState<State>();
 *
 * Before(async (state) => {
 *   return { ...state, counter: 0 };
 * });
 *
 * Given('the counter is {int}', async (state, [value]) => {
 *   return { ...state, counter: value };
 * });
 * ```
 *
 * @returns Hook and step definitions with state type inference
 */
export const withState = <State>(): WithState<State> => {
  const addStepWithState: AddStepFunctionWithState<State> = (pattern, fn) => {
    addStep(pattern, fn);
  };

  return {
    BeforeAll: BeforeAll<State>,
    Before: Before<State>,
    BeforeStep: BeforeStep<State>,
    AfterStep: AfterStep<State>,
    After: After<State>,
    AfterAll: AfterAll<State>,
    Given: addStepWithState,
    When: addStepWithState,
    Then: addStepWithState,
  };
};

export const BeforeAll: AddHookFunction<unknown> = (...args: unknown[]) => {
  addHook('beforeAll', ...args);
};

export const Before: AddHookFunction<unknown> = (...args: unknown[]) => {
  addHook('before', ...args);
};

export const BeforeStep: AddHookFunction<unknown> = (...args: unknown[]) => {
  addHook('beforeStep', ...args);
};

export const AfterStep: AddHookFunction<unknown> = (...args: unknown[]) => {
  addHook('afterStep', ...args);
};

export const After: AddHookFunction<unknown> = (...args: unknown[]) => {
  addHook('after', ...args);
};

export const AfterAll: AddHookFunction<unknown> = (...args: unknown[]) => {
  addHook('afterAll', ...args);
};

export const Given: AddStepFunction<unknown> = (pattern, fn) => {
  addStep(pattern, fn);
};

export const When: AddStepFunction<unknown> = (pattern, fn) => {
  addStep(pattern, fn);
};
export const Then: AddStepFunction<unknown> = (pattern, fn) => {
  addStep(pattern, fn);
};

/**
 * Extracts and transforms the data table from a PickleStep argument
 *
 * @example
 * ```gherkin
 * Given I have the following data:
 *   | Name  | Age |
 *   | Alice | 30  |
 *   | Bob   | 25  |
 * ```
 *
 * The step definition can use `getDataTable` to convert the argument into a more usable format:
 *
 * ```ts
 * import { getDataTable, Given } from '@aboviq/bun-test-cucumber';
 *
 * Given('I have the following data:', async (state, _args, data) => {
 *   const table = getDataTable(data);
 *   console.log(table);
 *   // Output:
 *   // [
 *   //   { Name: 'Alice', Age: '30' },
 *   //   { Name: 'Bob', Age: '25' }
 *   // ]
 *   return state;
 * });
 * ```
 * @param argument The argument from a PickleStep (third parameter of a step function)
 * @returns An array of records representing the data table
 */
export const getDataTable = (argument: PickleStepArgument | undefined): Record<string, string>[] => {
  if (!argument || !argument.dataTable) {
    throw new Error('No data table argument provided');
  }

  const [headerRow, ...rows] = argument.dataTable.rows;

  if (!headerRow) {
    return [];
  }

  const headers = headerRow.cells.map((cell) => cell.value);

  return rows.map((row) => {
    const record: Record<string, string> = {};
    row.cells.forEach((cell, index) => {
      const header = headers[index] ?? `Column ${index + 1}`;
      record[header] = cell.value;
    });
    return record;
  });
};
