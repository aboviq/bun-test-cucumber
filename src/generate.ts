import { resolveSync } from 'bun';

import type { GherkinDocument, Pickle } from '@cucumber/messages';
import parseTags from '@cucumber/tag-expressions';

import { escape } from './util';

type TagExpression = ReturnType<typeof parseTags>;

const getPickleName = (pickle: Pickle, index: number, pickles: readonly Pickle[]): string => {
  const name = escape(pickle.name);

  const indexesWithSameName: number[] = [];

  pickles.forEach((p, i) => {
    if (p.name === pickle.name) {
      indexesWithSameName.push(i);
    }
  });

  if (indexesWithSameName.length > 1) {
    return `${index + 1 - (indexesWithSameName[0] ?? 0)}. ${escape(name)}`;
  }

  return escape(name);
};

const getSkip = (pickle: Pickle, tagExpression?: TagExpression): string => {
  if (!tagExpression) {
    return '';
  }

  const tags = pickle.tags.map((tag) => tag.name);

  if (!tagExpression.evaluate(tags)) {
    return '.skip';
  }

  return '';
};

interface GenerateOptions {
  imports?: string;
  tagExpression?: TagExpression;
}

export const generate = (
  gherkinDocument: GherkinDocument,
  pickles: readonly Pickle[],
  { imports, tagExpression }: GenerateOptions = {},
): string => {
  const featureName = gherkinDocument.feature?.name ?? gherkinDocument.uri;

  if (!featureName) {
    throw new Error('Feature name is missing');
  }

  return `import { afterAll, afterEach, beforeAll, beforeEach, describe, it } from 'bun:test';
  import { applyHooks, runStep } from '${resolveSync('./index.ts', import.meta.dir)}';
  ${imports ?? ''}
  let state = {};
  
  beforeAll(async () => {
    state = await applyHooks('beforeAll', state);
  });
  
  afterAll(async () => {
    state = await applyHooks('afterAll', state);
  });

  describe('${escape(featureName)}', () => {
    ${pickles
      .map(
        (
          pickle,
          index,
        ) => `describe${getSkip(pickle, tagExpression)}('${getPickleName(pickle, index, pickles)}', () => {
      beforeAll(async () => {
        state = await applyHooks('before', state, ${JSON.stringify(pickle.tags)});
      });
      afterAll(async () => {
        state = await applyHooks('after', state, ${JSON.stringify(pickle.tags)});
      });
      beforeEach(async () => {
        state = await applyHooks('beforeStep', state, ${JSON.stringify(pickle.tags)});
      });
      afterEach(async () => {
        state = await applyHooks('afterStep', state, ${JSON.stringify(pickle.tags)});
      });
      ${pickle.steps
        .map(
          (step) => `it('${escape(step.text)}', async () => {
        state = await runStep(${JSON.stringify(step)}, state);
      });`,
        )
        .join('\n    ')}
    });`,
      )
      .join('\n  ')}
  });
`;
};
