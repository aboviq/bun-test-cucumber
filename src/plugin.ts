import { Glob, type BunPlugin } from 'bun';
import { readFileSync } from 'node:fs';

import { AstBuilder, compile, GherkinClassicTokenMatcher, Parser } from '@cucumber/gherkin';
import { IdGenerator, type GherkinDocument, type Pickle } from '@cucumber/messages';
import parseTags from '@cucumber/tag-expressions';

import { generate } from './generate';

interface BunTestCucumberOptions {
  /**
   * Current working directory to resolve the `stepDefinitionsPattern` from
   *
   * If not set, will use `process.cwd()`.
   *
   * This needs to be set if running `bun test` from a different directory than the project root,
   */
  cwd?: string;
  /**
   * Cucumber tag filter expression, e.g. `@smoke and not @wip`
   *
   * If not set, will use the `CUCUMBER_TAG_FILTER` environment variable if present.
   * If neither is set, all scenarios will be included.
   */
  tagFilter?: string;
  /**
   * Glob pattern to the step definitions files
   *
   * The pattern is relative to `cwd` (or process.cwd() if not set),
   * i.e. not automatically relative to the project root.
   */
  stepDefinitionsPattern: string;
  /**
   * Callback function to be called when generating the test files
   *
   * Can be used for logging or debugging purposes.
   *
   * @param gherkinDocument The parsed Gherkin document
   * @param pickles The generated pickles, i.e. scenarios
   * @param result The generated test file content
   */
  onGenerate?: (gherkinDocument: GherkinDocument, pickles: readonly Pickle[], result: string) => void;
}

export const bunTestCucumber = (options: BunTestCucumberOptions): BunPlugin => {
  return {
    name: '@aboviq/bun-test-cucumber',
    async setup(build) {
      const tagFilter = options.tagFilter ?? process.env['CUCUMBER_TAG_FILTER'];
      const tagExpression = tagFilter ? parseTags(tagFilter) : undefined;
      const stepDefinitionGlob = new Glob(options.stepDefinitionsPattern);

      let stepDefinitionsImports = '// Auto-imported step definitions:\n';

      for await (const file of stepDefinitionGlob.scan({
        cwd: options.cwd,
        absolute: true,
      })) {
        stepDefinitionsImports += `import "${file}";\n`;
      }

      build.onLoad({ filter: /\.feature$/ }, async (args) => {
        const uuidFn = IdGenerator.incrementing();
        const parser = new Parser(new AstBuilder(uuidFn), new GherkinClassicTokenMatcher());
        const content = readFileSync(args.path, 'utf-8');

        const gherkinDocument = parser.parse(content);
        const pickles = compile(gherkinDocument, args.path, uuidFn);

        const text = generate(gherkinDocument, pickles, {
          imports: stepDefinitionsImports,
          tagExpression,
        });

        options.onGenerate?.(gherkinDocument, pickles, text);

        return {
          contents: text,
          loader: 'ts',
        };
      });
    },
  };
};
