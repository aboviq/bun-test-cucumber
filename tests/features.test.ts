import { loadFeatures } from '../src';

await loadFeatures('features/**/*.feature');
