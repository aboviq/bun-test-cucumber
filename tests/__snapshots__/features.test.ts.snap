// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`My Feature with Tags Some Scenario with array data I have data 1`] = `
[
  "array",
  "before",
  "first",
  "second",
]
`;

exports[`My Feature with Tags Some Scenario with another array of data I have data 1`] = `
[
  "array",
  "when",
]
`;

exports[`My Feature with Outline 1. Some Scenario with Outline and big integers It should match snapshot 1`] = `1n`;

exports[`My Feature with Outline 2. Some Scenario with Outline and big integers It should match snapshot 1`] = `2n`;

exports[`My Feature with Outline 1. Some Scenario with Outline and integers It should match snapshot 1`] = `3`;

exports[`My Feature with Outline 2. Some Scenario with Outline and integers It should match snapshot 1`] = `5`;

exports[`With Data With Data Scenario I have data 1`] = `
[
  {
    "Header 1": "Value 1",
    "Header 2": "Value 2",
    "Header 3": "Value 3",
  },
  {
    "Header 1": "Value A",
    "Header 2": "Value B",
    "Header 3": "Value C",
  },
]
`;
